chainindex/indexer.go

```diff
@@ -0,0 +1,475 @@
+	if err != nil {
+		return xerrors.Errorf("failed to index eth tx hash: %w", err)
+	}
+
+	return nil
+}
+
+func (si *SqliteIndexer) Apply(ctx context.Context, from, to *types.TipSet) error {
+	si.closeLk.RLock()
+	if si.closed {
+		return ErrClosed
+	}
+	si.closeLk.RUnlock()
+
+	// We're moving the chain ahead from the `from` tipset to the `to` tipset
+	// Height(to) > Height(from)
+	err := withTx(ctx, si.db, func(tx *sql.Tx) error {
+		// index the `to` tipset first as we only need to index the tipsets and messages for it
+		if err := si.indexTipset(ctx, tx, to); err != nil {
+			return xerrors.Errorf("error indexing tipset: %w", err)
+		}
+
+		// index the `from` tipset just in case it's not indexed
+		if err := si.indexTipset(ctx, tx, from); err != nil {
+			return xerrors.Errorf("error indexing tipset: %w", err)
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return xerrors.Errorf("error applying tipset: %w", err)
+	}
+
+	si.notifyUpdateSubs()
+
+	return nil
+}
+
+func (si *SqliteIndexer) Revert(ctx context.Context, from, to *types.TipSet) error {
+	si.closeLk.RLock()
+	if si.closed {
+		return ErrClosed
+	}
+	si.closeLk.RUnlock()
+
+	// We're reverting the chain from the tipset at `from` to the tipset at `to`.
+	// Height(to) < Height(from)
+
+	revertTsKeyCid, err := toTipsetKeyCidBytes(from)
+	if err != nil {
+		return xerrors.Errorf("error getting tipset key cid: %w", err)
+	}
+
+	err = withTx(ctx, si.db, func(tx *sql.Tx) error {
+		if _, err := tx.Stmt(si.revertTipsetStmt).ExecContext(ctx, revertTsKeyCid); err != nil {
+			return xerrors.Errorf("error marking tipset as reverted: %w", err)
+		}
+
+		// index the `to` tipset as it has now been applied -> simply for redundancy
+		if err := si.indexTipset(ctx, tx, to); err != nil {
+			return xerrors.Errorf("error indexing tipset: %w", err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return xerrors.Errorf("error during revert transaction: %w", err)
+	}
+
+	si.notifyUpdateSubs()
+
+	return nil
+}
+
+func (si *SqliteIndexer) indexTipset(ctx context.Context, tx *sql.Tx, ts *types.TipSet) error {
+	tsKeyCidBytes, err := toTipsetKeyCidBytes(ts)
+	if err != nil {
+		return xerrors.Errorf("error computing tipset cid: %w", err)
+	}
+
+	restored, err := si.restoreTipsetIfExists(ctx, tx, tsKeyCidBytes)
+	if err != nil {
+		return xerrors.Errorf("error restoring tipset: %w", err)
+	}
+	if restored {
+		return nil
+	}
+
+	height := ts.Height()
+	insertTipsetMsgStmt := tx.Stmt(si.insertTipsetMessageStmt)
+
+	msgs, err := si.cs.MessagesForTipset(ctx, ts)
+	if err != nil {
+		return xerrors.Errorf("error getting messages for tipset: %w", err)
+	}
+
+	if len(msgs) == 0 {
+		// If there are no messages, just insert the tipset and return
+		if _, err := insertTipsetMsgStmt.ExecContext(ctx, tsKeyCidBytes, height, 0, nil, -1); err != nil {
+			return xerrors.Errorf("error inserting empty tipset: %w", err)
+		}
+		return nil
+	}
+
+	for i, msg := range msgs {
+		msg := msg
```

Comment:
It's just a safety habit I have when iterating over pointer variables in Golang to avoid the "new pointer var is pointing to same address" problem while looping.